In [43]:
import pandas as pd
import numpy as np

df = pd.read_csv("spam.csv")

Spam.csv ""

BAG OF WORDS REPRESENTATION USING NLP AND MACHINE LEARNING ALGORITHMS

In [44]:
df.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [45]:
df.Category.value_counts()

Category
ham     4825
spam     747
Name: count, dtype: int64

In [46]:
df.shape

(5572, 2)

In [47]:
def get_spam(x):
    if x == 'spam':
        return 1
    else:
        return 0

In [48]:
df["Spam"] = df["Category"].apply(get_spam)

we can apply also lambda x: 1 if x=='spam' else 0

In [49]:
df.head()

Category                                            Message  Spam
0      ham  Go until jurong point, crazy.. Available only ...     0
1      ham                      Ok lar... Joking wif u oni...     0
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...     1
3      ham  U dun say so early hor... U c already then say...     0
4      ham  Nah I don't think he goes to usf, he lives aro...     0

In [50]:
from sklearn.model_selection import train_test_split

X_train , X_test , y_train , y_test = train_test_split(df.Message , df.Spam , test_size=0.2)

In [51]:
X_train.shape #TRAINING DATA

(4457,)

In [52]:
X_test.shape

(1115,)

In [53]:
X_train[:4]

3480    Be sure to check your yahoo email. We sent pho...
4358    HELLOGORGEOUS, HOWS U? MY FONE WAS ON CHARGE L...
5320                 But we havent got da topic yet rite?
3810              At the funeral home with Audrey and dad
Name: Message, dtype: object

TEXT REPRESENTATION USING BoW

In [54]:
from sklearn.feature_extraction.text import CountVectorizer
v = CountVectorizer()

In [55]:
X_train_cv = v.fit_transform(X_train.values)
X_train_cv

<4457x7712 sparse matrix of type '<class 'numpy.int64'>'
	with 58955 stored elements in Compressed Sparse Row format>

CLASSIFY BAG OF WORDS AS TO MAKE ARRAY 

In [56]:
X_train_cv.toarray()[:2][0]

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [57]:
X_train_cv.shape

(4457, 7712)

in 4457 data we have 7786 unique words and v.vocabulary_ use to present all vocabs

In [71]:
v.get_feature_names_out()[2550]

'email'

In [59]:
X_train_np = X_train_cv.toarray()
X_train_np[0]

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [61]:
np.where(X_train_np[0]!=0)

(array([1261, 1742, 2550, 5158, 5991, 6599, 6905, 7383, 7635, 7658, 7677],
       dtype=int64),)

In [68]:
X_train[:4][3480]

'Be sure to check your yahoo email. We sent photos yesterday'

Now Build our ML model for prediction

In [73]:
from sklearn.naive_bayes import MultinomialNB

In [74]:
model = MultinomialNB()
model.fit(X_train_cv , y_train)

MultinomialNB()

In [77]:
X_test_cv = v.transform(X_test)

In [79]:
from sklearn.metrics import classification_report
y_pred = model.predict(X_test_cv)
print(classification_report(y_test , y_pred))

              precision    recall  f1-score   support

           0       0.99      1.00      0.99       966
           1       0.97      0.95      0.96       149

    accuracy                           0.99      1115
   macro avg       0.98      0.97      0.98      1115
weighted avg       0.99      0.99      0.99      1115



In [84]:
model.predict(X_train_cv)[3480]

0

For checking the spam or not spam emails do

In [87]:
emails = [
    'Upto 20% discount on parkiing , exclusive offer just for you. Dont miss this reward!' ,
    'Hey mohan, can we get together to watch footbal game tomorrow' 
    
]
emails_count = v.transform(emails)
model.predict(emails_count)

array([1, 0], dtype=int64)